20240927-multiple-object-deletion.md

```diff
@@ -0,0 +1,174 @@
+
+### Goals
+
+The goal of the implementation described by this document is to allow multiple objects to be deleted in a single request to the satellite.
+
+### Approach / Design
+
+#### Satellite
+
+The satellite metainfo RPC service must be extended with an endpoint allowing for the deletion of multiple objects in a single request.
+
+Requests directed to this endpoint must contain the following information:
+
+- The requester's API key
+- The bucket from which objects will be deleted
+- An indication of whether governance mode Object Lock restrictions should be bypassed
+- An indication of whether quiet mode should be enabled.
+  - With quiet mode enabled, only errors are returned; information about successful deletions is omitted from responses. Supporting this mode would allow us to save bandwidth on clients that use it.
+- A list of up to 1000 objects to delete
+
+The endpoint must respond with the following information:
+
+- A list containing the details of each successful deletion, including:
+  - The deleted object's key
+  - The deleted object's version (if the version wasn't specified in the request)
+  - The version ID of the delete marker (if applicable)
+- A list containing the details of each failed deletion, including:
+  - An error code
+  - The object's key
+  - The object's version ID (if it was provided in the request)
+
+Each bulk object deletion request will count as a single operation against the requester's deletion rate limit as opposed to each individual object deletion being counted separately.
+
+#### Libuplink
+
+Libuplink must provide a method through which the satellite metainfo RPC service's multiple object deletion endpoint is contacted. This is necessary for the gateway to function.
+
+#### Gateway
+
+The gateway must be given a configuration option that indicates whether the new multiple object deletion feature should be used. When it is enabled, it must invoke the new Libuplink method.
+
+#### Protobuf
+
+The protobuf definition for the metainfo RPC service must include a method for deleting multiple objects. Additionally, request and response types must be defined.
+
+```proto
+service Metainfo {
+    rpc DeleteObjects(DeleteObjectsRequest) returns (DeleteObjectsResponse);
+}
+
+message DeleteObjectsRequest {
+    RequestHeader header = 15;
+
+    bytes bucket = 1;
+    bool bypass_governance_retention = 2;
+    bool quiet = 3;
+    repeated DeleteObjectsRequestItem items = 4;
+}
+
+message DeleteObjectsRequestItem {
+    bytes encrypted_object_key = 1;
+    bytes object_version = 2;
+}
+
+message DeleteObjectsResponse {
+    repeated DeleteObjectsResponseDeletedItem deleted = 1;
```

Comment:
would it make sense to have a single repeated delete struct that also contains optional errors? that may fit better with the way S3 structures their response and could make our implementation simplier. Reference: https://docs.aws.amazon.com/AmazonS3/latest/API/API_DeleteObjects.html

e.g. `repeated DeleteResultItem items = 1;`

`message DeleteResultItem {
    bytes encrypted_object_key = 1;
    bytes deleted_object_version = 2;
    bytes delete_marker_version = 3;
    int32 error_code = 1;
}

Reply:
oh, two different structures seems like it would fit better. I'll mark this as resolved.